File: backend/app/utils/country.py
"""Country name normalization and matching for discovery and network filters."""

from __future__ import annotations

COUNTRY_ALIASES: dict[str, str] = {
    "usa": "United States",
    "us": "United States",
    "u.s.": "United States",
    "u.s.a.": "United States",
    "united states of america": "United States",
    "uk": "United Kingdom",
    "u.k.": "United Kingdom",
    "great britain": "United Kingdom",
    "uae": "United Arab Emirates",
}


def normalize_country(country: str | None) -> str | None:
    if not country or not str(country).strip():
        return None
    raw = str(country).strip()
    return COUNTRY_ALIASES.get(raw.lower(), raw)


def countries_match(user_country: str | None, filter_value: str | None) -> bool:
    if not filter_value or not str(filter_value).strip():
        return True
    if not user_country or not str(user_country).strip():
        return False

    filter_norm = normalize_country(filter_value)
    user_norm = normalize_country(user_country)
    if not filter_norm or not user_norm:
        return False

    filter_lower = filter_norm.lower()
    user_lower = user_norm.lower()
    filter_raw = str(filter_value).strip().lower()
    user_raw = str(user_country).strip().lower()

    if user_lower == filter_lower:
        return True
    if filter_lower in user_lower or user_lower in filter_lower:
        return True
    if filter_raw in user_raw or user_raw in filter_raw:
        return True
    return False

File: frontend/lib/utils/country.ts
const COUNTRY_ALIASES: Record<string, string> = {
  usa: 'United States',
  us: 'United States',
  'u.s.': 'United States',
  'u.s.a.': 'United States',
  'united states of america': 'United States',
  uk: 'United Kingdom',
  'u.k.': 'United Kingdom',
  'great britain': 'United Kingdom',
  uae: 'United Arab Emirates',
}

export function normalizeCountryName(value: string): string {
  const trimmed = value.trim()
  if (!trimmed) return ''
  return COUNTRY_ALIASES[trimmed.toLowerCase()] ?? trimmed
}

/** Returns true when a user's country matches a search/filter value (aliases, partial, case-insensitive). */
export function countriesMatch(userCountry: string | null | undefined, filter: string): boolean {
  if (!filter.trim()) return true
  if (!userCountry?.trim()) return false

  const filterNorm = normalizeCountryName(filter).toLowerCase()
  const userNorm = normalizeCountryName(userCountry).toLowerCase()
  const filterRaw = filter.trim().toLowerCase()
  const userRaw = userCountry.trim().toLowerCase()

  if (userNorm === filterNorm) return true
  if (userNorm.includes(filterNorm) || filterNorm.includes(userNorm)) return true
  if (userRaw.includes(filterRaw) || filterRaw.includes(userRaw)) return true
  return false
}

export function getCountryFromSearchParams(search: string): string | null {
  if (!search) return null
  const value = new URLSearchParams(search).get('country')
  return value?.trim() || null
}
